### PR TITLE
[WIP] MSC0000: Space Flair

### PR DESCRIPTION
**Pull request by turt2live**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/travis/msc/space-flair/proposals/3219-space-flair.md)

### CODE REVIEW

File: proposals/3219-space-flair.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# MSC3219: Space Flair
```
Comment (ShadowJonathan) [inline proposals/3219-space-flair.md:1]: Bikeshed: Should the client repr of these flairs be "square-ish" as how they're currently displayed in Element's space sidebar?

Comment (t3chguy) [reply proposals/3219-space-flair.md:1]: Given that MSC1772 didn't say anything about square avatars that definitely seems like a client decision and off-topic for this MSC.

File: proposals/3219-space-flair.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+exposing only a subset of state to non-members is considered functionality for a future/different MSC.
+
+Clients should additionally be aware that Spaces can be peekable (history set to `world_readable`) but *not* joinable (join rules of 
+`invite` or similar). This can also represent how a community can create a concept of "official" members by inviting them to a peekable, 
+but non-public, space. Flair would be enabled for this Space to denote official community members.
+
+## Unstable prefix
+
+While this MSC is not considered stable, implementations should use `org.matrix.msc3219` as a namespace. This means `m.room.flair` 
+becomes `org.matrix.msc0000.room.flair` and `m.flair` becomes `org.matrix.msc3219.flair` for example.
```
Comment (Cadair) [inline proposals/3219-space-flair.md:74]:
```suggestion
becomes `org.matrix.msc3219.room.flair` and `m.flair` becomes `org.matrix.msc3219.flair` for example.
```

Comment (turt2live) [reply proposals/3219-space-flair.md:74]: Damn, knew I missed one. Thanks

File: proposals/3219-space-flair.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    "spaces": [
+        {"room": "!room:example.org", "via": ["matrix.org", "example.org"]},
+        {"room": "#alias:example.org"},
+    ]
+}
+```
+
+`via` is optional but recommended for routing purposes. The `room` can be a room ID or alias. Clients would peek these rooms to 
+get their avatar/aesthetic state events for representation. Clients are recommended to only show flair for rooms which are actually 
+spaces, though this proposal doesn't impose any limitations in rooms enabling flair of other non-space rooms.
```
Comment (kevincox) [inline proposals/3219-space-flair.md:26]: What is the benefit of ignoring non-space rooms?

File: proposals/3219-space-flair.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+Flair is disabled by default for all users. Individual members can set a `"m.flair": true` flag on their own membership event in 
+the relevant space to enable visibility of their flair. Clients should ignore this flag on non-`join` membership events to prevent 
+inviters, moderators, etc from "enabling" flair for the user without being an active participant. Servers are expected to preserve 
+the flag, if present, when changing profile information for the user (displayname/avatar). When not explicitly `true`, flair should 
+be considered disabled.
+
+## Alternatives
+
+### Extensible profiles ([MSC1769](https://github.com/matrix-org/matrix-doc/issues/1769))
```
Comment (kevincox) [inline proposals/3219-space-flair.md:40]: I think flair should be managed by-profile. This today can be accomplished by updating the `m.room.member` event and in the future by room-based-profiles. This solves the issue for a user managing which flair shows in which rooms.

File: proposals/3219-space-flair.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+with empty state key. The `content` would look similar to:
+
+```json5
+{
+    "spaces": [
+        {"room": "!room:example.org", "via": ["matrix.org", "example.org"]},
+        {"room": "#alias:example.org"},
+    ]
+}
+```
```
Comment (kevincox) [inline proposals/3219-space-flair.md:22]: This seems somewhat odd to me. A room doesn't have flair, why are we managing it in the room. I think the better option is to control flair in the user profiles, but the room can have a list of allowed flair (possibly open).

**Review by kevincox — COMMENTED** [review]
I don't like this. It seems like a quick hack to get it working but it seems like it will need to be redone as we get some required changes in.

- We need good profile support for users to manage what flair they show where. Putting the enable/disable in spaces forces this to be global.
- This has no way to allow arbitrary flair, the allowlist is set in the room.

TL;DR I see fair as a profile feature, the room/space can add some lists of allowed flair if they want.

File: proposals/3219-space-flair.md (hunk 4 above)
Comment (aaronraimist) [reply proposals/3219-space-flair.md:40]: > This today can be accomplished by updating the m.room.member event and in the future by room-based-profiles.

Isn't that exactly what this MSC is proposing?

From the MSC:
> Individual members can set a "m.flair": true flag on their own membership event in the relevant space to enable visibility of their flair.

Comment (kevincox) [reply proposals/3219-space-flair.md:40]: The key difference is that (if I understand correctly) this sets the enable flag on the **space**. Whereas the current profiles and most if not all MSCs about profiles put the config in the **room**. This means that with this MSC showing flair is a global decision per-flair (which is then filtered by the per-room whitelist, but the user is not in control there) whereas the other profiles (and profile proposals) allow you to decide how you want to appear in each room.

**Comment by marcusmueller** [comment]
Just a user/element renter's feedback: This feature was valuable for conference organizing in with Communities in the past, and it would be very cool to see it for spaces in the future.

File: proposals/3219-space-flair.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+`via` is optional but recommended for routing purposes. The `room` can be a room ID or alias. Clients would peek these rooms to 
+get their avatar/aesthetic state events for representation. Clients are recommended to only show flair for rooms which are actually 
+spaces, though this proposal doesn't impose any limitations in rooms enabling flair of other non-space rooms.
+
+This approach also allows rooms to enable flair for private spaces if the user viewing the flair is in the private space, though 
+this is not seen as a realistic usage scenario given users joined to the space would be able to identify the other space members 
+(typically).
+
+Flair is disabled by default for all users. Individual members can set a `"m.flair": true` flag on their own membership event in
```
Comment (HarHarLinks) [inline proposals/3219-space-flair.md:32]: As kevincox pointed out in other places, this way to manage it seems odd. Instead, the membership event should contain a list of flairs to display in the respective room.

- Would I globally want to enable my #community:archlinux.org flair to tell everybody that sees me "I use Arch BTW"?
- Why would I enable it for rooms in the Arch space? I'm in an Arch chatroom, obviously I use Arch BTW
- I image this would set a standard for rooms to disable badges by default, as nobody wants to see unrelated badges. This makes it seem better to design it the other way round: explicitly enable instead of explicitly block

**Review by HarHarLinks — CHANGES_REQUESTED** [review]
<sup>oops, can't seem to delete this</sup>

File: proposals/3219-space-flair.md (hunk 6 above)
Comment (HarHarLinks) [inline proposals/3219-space-flair.md:32]: As kevincox pointed out in other places, this way to manage it seems odd. Instead, the membership event should contain a list of flairs to display in the respective room.

- Would I globally want to enable my #community:archlinux.org flair to tell everybody that sees me "I use Arch BTW"?
- Why would I enable it for rooms in the Arch space? I'm in an Arch chatroom, obviously I use Arch BTW
- I imagine this would set a standard for rooms to disable badges by default, as nobody wants to see unrelated badges. This makes it seem better to design it the other way round: explicitly enable instead of explicitly block

File: proposals/3219-space-flair.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# MSC3219: Space Flair
+
+[Spaces](https://github.com/matrix-org/matrix-doc/pull/1772) introduced an improved way of organizing rooms into tree-like 
+structures in Matrix. Its predecessor, Groups, had a concept of "flair" (or "related groups" at the technical level) where 
+users could have a smaller version of the Group's avatar next to their name in messages they send, if they've enabled flair.
```
Comment (HarHarLinks) [inline proposals/3219-space-flair.md:5]: I feel like this could be much better and flexible instead of more or less copying old community flairs, in which I don't see too much benefit anyway.

Flairs should not be limited to spaces (or plain rooms):
Much more interesting things can be done with them, e.g.
- show off my matrix projects (and their rooms) in the matrix community space
- indicate who is an official team member (of a space/room), with access restricted badges through e.g. room power levels
- indicate who is in a (sub)team of a room that doesn't exist
- misc stuff, like "5 year club", that similarly has no extra rooms

These multiple badges could be modeled completely by room state. Alternatively by flairs-as-rooms as the proposal already does. Needed to add on: a room power level that defines the required PL to use the room as avatar (-> public rooms with flair PL <= default PL means anyone can). Additional flairs are created by creating new rooms with a `m.flair` flag in their state to indicate not showing these rooms in the room list. Again access management by PL. The flair rooms are announced in the "owner room's" state (is it a space yet? 🤔).

I could probably go on but I hope what I mean is evident? Imo this should be done properly right away and not have these additional kinds of flair tagged on later, potentially changing flairs again to not make a mess of the spec.

**Review by HarHarLinks — CHANGES_REQUESTED** [review]
Apparently I don't know how to review, here I try again.

I feel like flairs should be designed more general and flexible. This MSC seems to try to mirror the old community flairs. Essentially, it enables you to stick a couple badges on your belt to advertise for the spaces you're joined to where ever you go, unless you happen to go into a room that asks you to show only certain badges.

**Comment by HarHarLinks** [comment]
since this is closely related, instead of adding its own spec for flairs, simply put flairs into per-room/space [extensible profiles](https://github.com/matrix-org/matrix-spec-proposals/pull/1769) and which for example are attached to the respective rooms like [this](https://github.com/matrix-org/matrix-spec-proposals/pull/3755#discussion_r835901536)
